### DISCUSSION TITLE
[Azure Devops][Pipelines] Pomiar czasu na pieline.

### DISCUSSION THREAD

**Discussion by galante13** [AzureCommunityPL/group#88] [Pomoc] [answered]
Czy ktoś z Was może wie jak mierzyć czas wykonywania piepline i stworzyć mechanizm reagowania na spike-i?

**Comment by kamil-mrzyglod** [comment, marked as answer]
Wykorzystać REST API - https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/list?view=azure-devops-rest-6.1#build.

Co do reagowania - zależy co masz na myśli.

**Reply by galante13** [reply]
Może to być mail, jakaś notyfikacja na teamsach, możliwe, że logic app by spełniał potrzebę w takim razie (?).

**Reply by kamil-mrzyglod** [reply]
W takim razie cokolwiek co potrafi odpytać REST API i zinterpretować wyniki się nada. Czy to będzie Azure Function, Logic App, niezalezna appka to pewnie już sprawa drugorzędna.

**Reply by galante13** [reply]
Dzięki, już sobie zaprojektowałem rozwiązanie w okół tego api :)

**Comment by wi5nia** [comment]
Mozna bardzo ladnie do tego wykorzystac [Audit stream](https://docs.microsoft.com/en-us/azure/devops/organizations/audit/auditing-streaming?view=azure-devops) i sobie to wrzucic dalej i analizowac.

**Reply by kamil-mrzyglod** [reply]
O a tego nie znałem 🛒

**Reply by galante13** [reply]
Niestety to rozwiązanie nie dostarcza informacji o wywołaniach pipelin-ów.

**Reply by wi5nia** [reply]
Faktycznie, tylko dla releasow nie pipelinow

**Reply by galante13** [reply]
Dzięki anyway, czegoś się nauczyłem :)

**Reply by KamilWiecek** [reply]
Service hooki pozwoliłyby na zebranie tych i innych ciekawych danych. Można wysyłać na jakiś endpoint http, kolejkować na service bus, storage account queue, wrzucić na Teamsy itd:
https://docs.microsoft.com/en-us/azure/devops/service-hooks/overview?view=azure-devops
Tutaj fajny przykład z sentinelem:
https://medium.com/@anderseide/integrate-azure-devops-project-events-with-azure-sentinel-a58c2fb069d2

**Comment by JanuszNowak** [comment]
@galante13  Masz dostępna cała analitykę https://docs.microsoft.com/en-us/azure/devops/pipelines/reports/pipelinereport?view=azure-devops
Która jest dostępna tez przez api. Zależnie jak chcesz reagować możesz iterować po wszystkich bildach dla wszystkich pielinów lub tez dodać ktok która sprawdza czy wykonanie mieści się 10% poprzedniego rana i jak nie to tworzyć wokitem lub wykonąc inną akcje.

**Reply by galante13** [reply]
Przejrzałem azure devops services REST API i nie ma endpointu do gotowego analytics, może źle patrzę?

**Reply by kmadof** [reply]
Endpoint'u takiego co Ci zwróci na gotowe nie ma. Jest za to to konrektnego pipeline run czy build. Tutaj masz jeszcze coś co może Ci pomoże ogarnąć temat
https://stackoverflow.com/questions/64362196/azure-devops-api-discrepancy-between-run-and-build-start-times
